File: lib/cuite__models.ml
external get_payload
  : [>`Payload of 'a qt] -> 'a
  = "cuite_get_payload" [@@noalloc]

(* QAbstractTableModel *)
type 'a qOCamlTableModel =
  [ `QOCamlTableModel | `Payload of 'a | qAbstractTableModel ]

type 'a qOCamlTableModel'callback = {
  row_count    : 'a -> 'a qOCamlTableModel qt -> qModelIndex qt -> int;
  column_count : 'a -> 'a qOCamlTableModel qt -> qModelIndex qt -> int;
  data         : 'a -> 'a qOCamlTableModel qt -> qModelIndex qt -> qt'ItemDataRole -> QVariant.t;
  header_data  : 'a -> 'a qOCamlTableModel qt -> int -> qt'Orientation -> qt'ItemDataRole -> QVariant.t;
  flags        : 'a -> 'a qOCamlTableModel qt -> qModelIndex qt -> qt'ItemFlags;
  set_data     : 'a -> 'a qOCamlTableModel qt -> qModelIndex qt -> QVariant.t -> qt'ItemDataRole -> bool;
  insert_rows  : 'a -> 'a qOCamlTableModel qt -> int -> int -> qModelIndex qt -> (unit -> unit) option;
  remove_rows  : 'a -> 'a qOCamlTableModel qt -> int -> int -> qModelIndex qt -> (unit -> unit) option;
}

let qOCamlTableModel'callback
    ~row_count ~column_count ~data
    ?(header_data = fun _ _ _ _ _ -> QVariant.null)
    ?(flags = fun _ _ _ -> QFlags.empty)
    ?(set_data = fun _ _ _ _ _ -> false)
    ?(insert_rows = fun _ _ _ _ _ -> None)
    ?(remove_rows  = fun _ _ _ _ _ -> None)
    ()
  = { row_count ; column_count ;
      data ; header_data ; flags ; set_data ;
      insert_rows ; remove_rows }

external new'QOCamlTableModel : 'a qOCamlTableModel'callback -> 'a -> 'a qOCamlTableModel qt =
  "cuite_new_QOCamlTableModel"

(* QSyntaxHighlighter *)

type 'a qOCamlSyntaxHighlighter =
  [ `QOCamlSyntaxHighlighter | `Payload of 'a | qSyntaxHighlighter ]

type 'a qOCamlSyntaxHighlighter'callback = {
  highlight_block : 'a -> 'a qOCamlSyntaxHighlighter qt -> string -> unit;
  release_state : 'a -> 'a qOCamlSyntaxHighlighter qt -> int -> unit;
}

let qOCamlSyntaxHighlighter'callback ~highlight_block ~release_state =
  { highlight_block; release_state }

external new'QOCamlSyntaxHighlighter : 'a qOCamlSyntaxHighlighter'callback -> 'a ->
  [> qObject] qt option -> 'a qOCamlSyntaxHighlighter qt =
  "cuite_new_QOCamlSyntaxHighlighter"

module OCamlSyntaxHighlighter = struct
  external currentBlock : 'a qOCamlSyntaxHighlighter qt -> qTextBlock qt =
    "cuite_QOCamlSyntaxHighlighter_currentBlock"
  external currentBlockState : 'a qOCamlSyntaxHighlighter qt -> int =
    "cuite_QOCamlSyntaxHighlighter_currentBlockState"
  (*currentBlockUserData() const*)
  external format : 'a qOCamlSyntaxHighlighter qt -> int -> qTextCharFormat qt =
    "cuite_QOCamlSyntaxHighlighter_format"
  external previousBlockState : 'a qOCamlSyntaxHighlighter qt -> int =
    "cuite_QOCamlSyntaxHighlighter_previousBlockState"
  external setCurrentBlockState : 'a qOCamlSyntaxHighlighter qt -> int -> unit =
    "cuite_QOCamlSyntaxHighlighter_setCurrentBlockState"
  external setCurrentBlockStateAndNotifyRelease : 'a qOCamlSyntaxHighlighter qt -> int -> unit =
    "cuite_QOCamlSyntaxHighlighter_setCurrentBlockStateAndNotifyRelease"
      (*setCurrentBlockUserData(QTextBlockUserData *data)*)
  external setFormat : 'a qOCamlSyntaxHighlighter qt -> int -> int -> [> qTextCharFormat] qt -> unit =
    "cuite_QOCamlSyntaxHighlighter_setFormat"
  external setFormatColor : 'a qOCamlSyntaxHighlighter qt -> int -> int -> [> qColor] qt -> unit =
    "cuite_QOCamlSyntaxHighlighter_setFormatColor"
  external setFormatFont : 'a qOCamlSyntaxHighlighter qt -> int -> int -> [> qFont] qt -> unit =
    "cuite_QOCamlSyntaxHighlighter_setFormatFont"

end

module StatefulHighlighter(State : Hashtbl.HashedType) :
sig
  type t

  type context
  val get_format : context -> int -> qTextCharFormat qt
  val set_format : context -> int -> int -> [> qTextCharFormat] qt -> unit
  val set_format_color : context -> int -> int -> [> qColor] qt -> unit
  val set_format_font : context -> int -> int -> [> qFont] qt -> unit

  val make :
    State.t -> (context -> string -> State.t -> State.t) ->
    [> qObject] qt option -> qSyntaxHighlighter qt
end = struct
  module Table = Hashtbl.Make(State)

  type cell = {
    index: int;
    mutable users: int;
    state: State.t
  }

  type t = {
    mutable counter : int;
    index : cell Table.t;
    rev_index : (int, cell) Hashtbl.t
  }

  let acquire t state =
    match Table.find t.index state with
    | cell ->
      cell.users <- cell.users + 1;
      cell.index
    | exception Not_found ->
      let index = t.counter in
      t.counter <- t.counter + 1;
      let cell = { index; users = 1; state } in
      Table.add t.index state cell;
      Hashtbl.add t.rev_index index cell;
      index

  let release t _ index =
    match Hashtbl.find t.rev_index index with
    | exception Not_found ->
      prerr_endline "Cuite.StatefulHighlighter: missing index"
    | cell ->
      cell.users <- cell.users - 1;
      if cell.users = 0 then (
        Table.remove t.index cell.state;
        Hashtbl.remove t.rev_index cell.index
      ) else
        assert (cell.users > 0)

  type context = t qOCamlSyntaxHighlighter qt
  let get_format = OCamlSyntaxHighlighter.format
  let set_format = OCamlSyntaxHighlighter.setFormat
  let set_format_color = OCamlSyntaxHighlighter.setFormatColor
  let set_format_font = OCamlSyntaxHighlighter.setFormatFont

  let make initial f parent =
    let table = {
      counter = 0;
      index = Table.create 7;
      rev_index = Hashtbl.create 7;
    } in
    let result =
      new'QOCamlSyntaxHighlighter {
        highlight_block = (fun table qt text ->
            let state =
              match OCamlSyntaxHighlighter.previousBlockState qt with
              | -1 -> initial
              |  n -> (Hashtbl.find table.rev_index n).state
            in
            let state' : State.t = f qt text state in
            OCamlSyntaxHighlighter.setCurrentBlockStateAndNotifyRelease
              qt (acquire table state')
          );
        release_state = release;
      } table parent
    in
    (result : t qOCamlSyntaxHighlighter qt :> qSyntaxHighlighter qt)
end

(* QEventFilter *)

type 'a qOCamlEventFilter =
  [ `QOCamlEventFilter | `Payload of 'a | qObject ]

type 'a qOCamlEventFilter'callback = {
  filter : 'a -> 'a qOCamlEventFilter qt -> qObject qt -> qEvent qt -> bool;
}

let qOCamlEventFilter'callback ~filter =
  { filter }

external new'QOCamlEventFilter : 'a qOCamlEventFilter'callback -> 'a -> 'a qOCamlEventFilter qt =
  "cuite_new_QOCamlEventFilter"

File: cpp/cuite_models.cpp
#include <caml/memory.h>
#include <caml/callback.h>
#include <QAbstractTableModel>
#include <QSyntaxHighlighter>
#include <QtGui>
#include "cuite_support.h"
#include "cuite_wrappers.h"
#include "cuite_variant.h"
#include "flags.hpp.gen"

#define model_payload(v) Field(v, QObject_fields_count + 0)
#define model_callback(v, n) Field(Field(v, QObject_fields_count + 1), n)

#define ID(x) x

#define PUSH_ARG(v) *_call_top = (v); _call_top += 1

#define PREPARE_CALL(n)                            \
  {                                                \
    CUITE_OCAML_REGION();                          \
    value *_call_args = cuite_region_allocn((n)+1); \
    value *_call_top = _call_args

#define CALL_CLOSURE(result, inj, closure)                                        \
    *_call_top = caml_callbackN_exn(closure, _call_top - _call_args, _call_args); \
    if (Is_exception_result(*_call_top))                                          \
      cuite_debug_aborted_callback(__func__, Extract_exception(*_call_top));       \
    else                                                                          \
      result = inj(*_call_top);                                                   \
  }

#define PREPARE_METHOD(n, obj)                 \
  PREPARE_CALL((n)+2);                             \
    value& _call_this = cuite_QObject_to_ocaml(obj); \
    PUSH_ARG(model_payload(_call_this)); \
    PUSH_ARG(_call_this)

#define CALL_METHOD(result, inj, offset) \
  CALL_CLOSURE(result, inj, model_callback(_call_this, offset))

external value cuite_get_payload(value v)
{
  return model_payload(v);
}

/* QAbstractTableModel */

class QOCamlTableModel : public QAbstractTableModel
{
  public:
    QOCamlTableModel();

    int rowCount(const QModelIndex& parent) const override;
    int columnCount(const QModelIndex& parent) const override;
    QVariant data(const QModelIndex& index, int role) const override;
    QVariant headerData(int section, Qt::Orientation orientation, int role) const override;
    Qt::ItemFlags flags(const QModelIndex &index) const override;
    bool setData(const QModelIndex &index, const QVariant &v, int role = Qt::EditRole) override;
    bool insertRows(int position, int rows, const QModelIndex &index = QModelIndex()) override;
    bool removeRows(int position, int rows, const QModelIndex &index = QModelIndex()) override;
};

#define meth_row_count    0
#define meth_column_count 1
#define meth_data         2
#define meth_header_data  3
#define meth_flags        4
#define meth_set_data     5
#define meth_insert_rows  6
#define meth_remove_rows  7

QOCamlTableModel::QOCamlTableModel()
  : QAbstractTableModel()
{
}

external value cuite_new_QOCamlTableModel(value callbacks, value payload)
{
  CUITE_GC_REGION(&callbacks, &payload);

  QOCamlTableModel *model = new QOCamlTableModel();
  value& v(cuite_QObject_allocate_value(model, 2));
  Store_field(v, QObject_fields_count + 0, payload);
  Store_field(v, QObject_fields_count + 1, callbacks);
  CUITE_LOG("new QOCamlTableModel(callbacks = %08lx) = %08lx\n", callbacks, v);
  return v;
}

int QOCamlTableModel::rowCount(const QModelIndex& parent) const
{
  int result = 0;
  PREPARE_METHOD(1, this);
    PUSH_ARG(cuite_QModelIndex_to_ocaml(parent));
  CALL_METHOD(result, Long_val, meth_row_count);
  return result;
}

int QOCamlTableModel::columnCount(const QModelIndex& parent) const
{
  int result = 0;
  PREPARE_METHOD(1, this);
    PUSH_ARG(cuite_QModelIndex_to_ocaml(parent));
  CALL_METHOD(result, Long_val, meth_column_count);
  return result;
}

QVariant QOCamlTableModel::data(const QModelIndex& index, int role) const
{
  QVariant result;
  PREPARE_METHOD(2, this);
    PUSH_ARG(cuite_QModelIndex_to_ocaml(index));
    PUSH_ARG(cuite_Qt_ItemDataRole_to_ocaml((Qt::ItemDataRole)role));
  CALL_METHOD(result, cuite_QVariant_from_ocaml, meth_data);
  return result;
}

QVariant QOCamlTableModel::headerData(int section, Qt::Orientation orientation, int role) const
{
  QVariant result;
  PREPARE_METHOD(3, this);
    PUSH_ARG(Val_long(section));
    PUSH_ARG(cuite_Qt_Orientation_to_ocaml(orientation));
    PUSH_ARG(cuite_Qt_ItemDataRole_to_ocaml((Qt::ItemDataRole)role));
  CALL_METHOD(result, cuite_QVariant_from_ocaml, meth_header_data);
  return result;
}

static Qt::ItemFlags cuite_Qt_ItemFlags_from_ocaml(const value& v)
{
  int64_t x = cuite_flag_from_ocaml(v);
  return Qt::ItemFlags((int)x);
}

Qt::ItemFlags QOCamlTableModel::flags(const QModelIndex &index) const
{
  Qt::ItemFlags result = 0;
  PREPARE_METHOD(1, this);
    PUSH_ARG(cuite_QModelIndex_to_ocaml(index));
  CALL_METHOD(result, cuite_Qt_ItemFlags_from_ocaml, meth_flags);
  return result;
}

bool QOCamlTableModel::setData(const QModelIndex &index, const QVariant &v, int role)
{
  bool result = false;
  PREPARE_METHOD(3, this);
    PUSH_ARG(cuite_QModelIndex_to_ocaml(index));
    PUSH_ARG(cuite_QVariant_to_ocaml(&v));
    PUSH_ARG(cuite_Qt_ItemDataRole_to_ocaml((Qt::ItemDataRole)role));
  CALL_METHOD(result, Val_bool, meth_set_data);
  return result;
}

#define ID(x) x

bool QOCamlTableModel::insertRows(int position, int rows, const QModelIndex &index)
{
  value result = Val_unit;
  PREPARE_METHOD(3, this);
    PUSH_ARG(Val_long(position));
    PUSH_ARG(Val_long(rows));
    PUSH_ARG(cuite_QModelIndex_to_ocaml(index));
  CALL_METHOD(result, ID, meth_insert_rows);

  if (result == Val_unit)
    return false;

  beginInsertRows(index, position, position + rows - 1);
  PREPARE_CALL(1);
    PUSH_ARG(Val_unit);
  CALL_CLOSURE(result, ID, Field(result, 0));
  endInsertRows();

  return true;
}

bool QOCamlTableModel::removeRows(int position, int rows, const QModelIndex &index)
{
  value result = Val_unit;
  PREPARE_METHOD(3, this);
    PUSH_ARG(Val_long(position));
    PUSH_ARG(Val_long(rows));
    PUSH_ARG(cuite_QModelIndex_to_ocaml(index));
  CALL_METHOD(result, ID, meth_remove_rows);

  if (result == Val_unit)
    return false;

  beginRemoveRows(index, position, position + rows - 1);
  PREPARE_CALL(1);
    PUSH_ARG(Val_unit);
  CALL_CLOSURE(result, ID, Field(result, 0));
  endRemoveRows();

  return true;
}


/* QSyntaxHighlighter */

#define meth_highlight_block 0
#define meth_release_data    1

class QOCamlSyntaxHighlighter : public QSyntaxHighlighter
{
  public:
    QOCamlSyntaxHighlighter(QObject *parent);
    //QOCamlSyntaxHighlighter(QTextDocument *parent);

    inline QTextBlock call_currentBlock() const
    { return this->currentBlock(); }
    inline int call_currentBlockState() const
    { return this->currentBlockState(); }
    inline QTextBlockUserData* call_currentBlockUserData() const
    { return this->currentBlockUserData(); }
    inline QTextCharFormat call_format(int position) const
    { return this->format(position); }
    inline int call_previousBlockState() const
    { return this->previousBlockState(); }
    inline void call_setFormat(int start, int count, const QTextCharFormat &format)
    { this->setFormat(start, count, format); }
    inline void call_setFormat(int start, int count, const QColor &color)
    { this->setFormat(start, count, color); }
    inline void call_setFormat(int start, int count, const QFont &font)
    { this->setFormat(start, count, font); }

    inline void call_setCurrentBlockState(int newState)
    {
      this->setCurrentBlockState(newState);
      this->setCurrentBlockUserData(NULL);
    }

    struct QOCamlTextBlockUserData : QTextBlockUserData
    {
      QOCamlSyntaxHighlighter *highlighter;
      int state;

      QOCamlTextBlockUserData(QOCamlSyntaxHighlighter *_highlighter, int _state)
        : highlighter(_highlighter), state(_state)
      {}

      void changeState(int newState)
      {
        value result;
        PREPARE_METHOD(1, highlighter);
        PUSH_ARG(Val_long(state));
        CALL_METHOD(result, ID, meth_release_data);
        state = newState;
      }

      ~QOCamlTextBlockUserData()
      {
        changeState(-1);
      }
    };

    inline void call_setCurrentBlockStateAndNotifyRelease(int newState)
    {
      this->setCurrentBlockState(newState);
      QTextBlockUserData *userData = this->currentBlockUserData();
      QOCamlTextBlockUserData *ocamlData = NULL;
      if (userData != NULL)
        ocamlData = dynamic_cast<QOCamlTextBlockUserData*>(userData);
      if (ocamlData != NULL)
        ocamlData->changeState(newState);
      else
      {
        ocamlData = new QOCamlTextBlockUserData(this, newState);
        this->setCurrentBlockUserData(ocamlData);
      }
    }

  protected:
    void highlightBlock(const QString &text) override;
};

QOCamlSyntaxHighlighter::QOCamlSyntaxHighlighter(QObject *parent)
  : QSyntaxHighlighter(parent) {}

//QOCamlSyntaxHighlighter::QOCamlSyntaxHighlighter(QTextDocument *parent)
//  : QSyntaxHighlighter(parent) {}

external value cuite_new_QOCamlSyntaxHighlighter(value callbacks, value payload, value parent)
{
  CUITE_GC_REGION(&callbacks, &payload, &parent);

  QOCamlSyntaxHighlighter *model = new QOCamlSyntaxHighlighter(cuite_QObject_option_from_ocaml(parent));
  value& v(cuite_QObject_allocate_value(model, 2));
  Store_field(v, QObject_fields_count + 0, payload);
  Store_field(v, QObject_fields_count + 1, callbacks);
  CUITE_LOG("new QOCamlSyntaxHighlighter(callbacks = %08lx) = %08lx\n", callbacks, v);
  return v;
}

void QOCamlSyntaxHighlighter::highlightBlock(const QString &text)
{
  value result;
  PREPARE_METHOD(1, this);
    PUSH_ARG(cuite_QString_to_ocaml(text));
  CALL_METHOD(result, ID, meth_highlight_block);
}

external value cuite_QOCamlSyntaxHighlighter_currentBlock(value vself)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(vself));
  CUITE_GC_REGION(&vself);
  QOCamlSyntaxHighlighter *self = (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(vself);
  QTextBlock block(self->call_currentBlock());
  return cuite_QTextBlock_to_ocaml(block);
}

external value cuite_QOCamlSyntaxHighlighter_currentBlockState(value vself)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(vself));
  CUITE_GC_REGION(&vself);
  QOCamlSyntaxHighlighter *self = (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(vself);
  return Val_long(self->call_currentBlockState());
}

external value cuite_QOCamlSyntaxHighlighter_format(value vself, value position)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(vself));
  CUITE_GC_REGION(&vself);
  QOCamlSyntaxHighlighter *self = (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(vself);
  QTextCharFormat result(self->call_format(Long_val(position)));
  return cuite_QTextCharFormat_to_ocaml(result);
}

external value cuite_QOCamlSyntaxHighlighter_previousBlockState(value vself)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(vself));
  CUITE_GC_REGION(&vself);
  QOCamlSyntaxHighlighter *self = (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(vself);
  return Val_long(self->call_previousBlockState());
}

external value cuite_QOCamlSyntaxHighlighter_setCurrentBlockState(value vself, value newState)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(vself));
  CUITE_GC_REGION(&vself);
  QOCamlSyntaxHighlighter *self = (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(vself);
  self->call_setCurrentBlockState(Long_val(newState));
  return Val_unit;
}

external value cuite_QOCamlSyntaxHighlighter_setCurrentBlockStateAndNotifyRelease(value vself, value newState)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(vself));
  CUITE_GC_REGION(&vself);
  QOCamlSyntaxHighlighter *self = (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(vself);
  self->call_setCurrentBlockStateAndNotifyRelease(Long_val(newState));
  return Val_unit;
}

external value cuite_QOCamlSyntaxHighlighter_setFormat(value mlself, value start, value count, value mlformat)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(mlself));
  CUITE_GC_REGION(&mlself, &mlformat);
  QOCamlSyntaxHighlighter *self =
    (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(mlself);
  const QTextCharFormat& format = cuite_QTextCharFormat_from_ocaml(mlformat);
  self->call_setFormat(Long_val(start), Long_val(count), format);
  return Val_unit;
}

external value cuite_QOCamlSyntaxHighlighter_setFormatColor(value mlself, value start, value count, value mlcolor)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(mlself));
  CUITE_GC_REGION(&mlself, &mlcolor);
  QOCamlSyntaxHighlighter *self =
    (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(mlself);
  const QColor& color = cuite_QColor_from_ocaml(mlcolor);
  self->call_setFormat(Long_val(start), Long_val(count), color);
  return Val_unit;
}

external value cuite_QOCamlSyntaxHighlighter_setFormatFont(value mlself, value start, value count, value mlfont)
{
  CHECK_USE_AFTER_FREE(cuite_QObject_check_use(mlself));
  CUITE_GC_REGION(&mlself, &mlfont);
  QOCamlSyntaxHighlighter *self =
    (QOCamlSyntaxHighlighter *)cuite_QObject_from_ocaml(mlself);
  const QFont& font = cuite_QFont_from_ocaml(mlfont);
  self->call_setFormat(Long_val(start), Long_val(count), font);
  return Val_unit;
}

/* Event Filter */

class QOCamlEventFilter : public QObject
{
  public:
    QOCamlEventFilter() {};

    bool eventFilter(QObject *watched, QEvent *event);
};

#define meth_filter    0

external value cuite_new_QOCamlEventFilter(value callbacks, value payload)
{
  CUITE_GC_REGION(&callbacks, &payload);

  QOCamlEventFilter *filter = new QOCamlEventFilter();
  value& v(cuite_QObject_allocate_value(filter, 2));
  Store_field(v, QObject_fields_count + 0, payload);
  Store_field(v, QObject_fields_count + 1, callbacks);
  CUITE_LOG("new QOCamlEventFilter(callbacks = %08lx) = %08lx\n", callbacks, v);
  return v;
}

bool QOCamlEventFilter::eventFilter(QObject *watched, QEvent *event)
{
  bool result = false;
  PREPARE_METHOD(2, this);
    PUSH_ARG(cuite_QObject_to_ocaml(watched));
    PUSH_ARG(cuite_QEvent_to_ocaml(event));
  CALL_METHOD(result, Val_bool, meth_filter);
  return result;
}

